cdap-app-fabric/src/main/java/co/cask/cdap/internal/profile/AdminEventPublisher.java

```diff
@@ -0,0 +1,162 @@
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.common.service.RetryStrategy;
+import co.cask.cdap.data2.metadata.writer.MetadataMessage;
+import co.cask.cdap.internal.app.runtime.schedule.ProgramSchedule;
+import co.cask.cdap.proto.id.ApplicationId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProfileId;
+import co.cask.cdap.proto.id.ScheduleId;
+import co.cask.cdap.proto.id.TopicId;
+import com.google.common.base.Throwables;
+import com.google.gson.Gson;
+import com.google.gson.JsonNull;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class for publishing the profile metadata change request to tms
+ */
+public class AdminEventPublisher {
+
+  private static final Gson GSON = new Gson();
+
+  private final TopicId topic;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  AdminEventPublisher(CConfiguration cConf) {
+    this.topic = NamespaceId.SYSTEM.topic(cConf.get(Constants.Metadata.MESSAGING_TOPIC));
+    this.retryStrategy = RetryStrategies.timeLimit(
+      20, TimeUnit.SECONDS,
+      RetryStrategies.exponentialDelay(10, 200, TimeUnit.MILLISECONDS));
+  }
+
+  /**
+   * Update the profile metadata for the affected programs/schedules, the metadata of programs/schedules will be
+   * updated according to the closest level preference that contains the profile information.
+   *
+   * @param messagingContext message context to get the publisher to publish the message
+   * @param entityId the entity id to update metadata, all programs/schedules under this entity id will be affected
+   * @param profileId the profileId which get assigned
+   */
+  public void publishProfileAssignment(MessagingContext messagingContext, EntityId entityId, ProfileId profileId) {
+    publishMessage(messagingContext, entityId, MetadataMessage.Type.PROFILE_ASSIGNMENT, profileId);
+  }
+
+  /**
+   * Publish a message when a profile is unassigned to an entity. The metadata of the programs/schedules will be
+   * reindexed to the closest level preference that contains the profile information
+   *
+   * @param messagingContext message context to get the publisher to publish the message
+   * @param entityId the entity id to reindex metadata, all programs/schedules under this entity id will be affected
+   */
+  public void publishProfileUnAssignment(MessagingContext messagingContext, EntityId entityId) {
+    publishMessage(messagingContext, entityId, MetadataMessage.Type.PROFILE_UNASSIGNMENT);
+  }
+
+  /**
+   * Publish a message about app creation, the programs/schedules which get created will get updated with the profile
+   * metadata
+   *
+   * @param messagingContext message context to get the publisher to publish the message
+   * @param applicationId the app id that get created
+   */
+  public void publishAppCreation(MessagingContext messagingContext, ApplicationId applicationId) {
```

Comment:
i think it would be better if this took the full ApplicationSpecification as well

Reply:
Added